lang/java/reef-tang/tang/src/main/java/org/apache/reef/tang/implementation/types/ConstructorDefImpl.java

```diff
@@ -129,6 +130,48 @@ public boolean equals(final Object o) {
     return equalsIgnoreOrder((ConstructorDef<?>) o);
   }
 
+  public boolean equalsNew(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    int length = getArgs().length;
+    ConstructorDef<?> def = (ConstructorDef<?>) o;
+    if (length != def.getArgs().length) {
+      return false;
+    } else if (length == 0) {
+      return true;
+    } else if (length == 1) {
+      return getArgs()[0].getName().equals(def.getArgs()[0].getName());
+    } else {
+      return equalsIgnoreOrderUsingHashMap(def);
+    }
+  }
+
+  /**
+   * Check to see if two boundConstructors take indistinguishable arguments. If
+   * so (and they are in the same class), then this would lead to ambiguous
+   * injection targets, and we want to fail fast.
+   *
+   * @param def
+   * @return
+   */
+  private boolean equalsIgnoreOrderUsingHashMap(final ConstructorDef<?> def) {
+    HashMap map = new HashMap();
```

Comment:
What's the benefit of using HashMap instead of HashSet? There should be not much performance difference, but readability-wise set looks better, as we don't need values, only keys